feat: Find symbols in current scope

tagyieh opened the pull request:
This PR aims to find all symbols in the current scope. 

Since `{ }` (curly brackets) are the delimiters of scope, the considered scopes where `source_file` (root node), `category_declaration`, `asset_declaration` and `associations_declaration`.

Additionally, "symbols" were considered to be both user-defined strings to identify components of the language (e.g. assets, categories, etc.) and keywords (e.g. `extends`, `abstract`, etc.)

Closes: #28

tagyieh committed:
Merge branch 'Tobiky:master' into feat/find_symbols_in_current_scope

tagyieh committed:
feat: Find symbols in current scope function.

tagyieh committed:
Merge branch 'Tobiky:master' into feat/find_symbols_in_current_scope

tagyieh committed:
chore: Fixed comments.

tagyieh committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/find_symbols_in_current_scope' into feat/find_symbols_in_current_scope

tagyieh committed:
merge: Merged master

tagyieh committed:
merge: Fixed file.

tagyieh committed:
tests: Find symbols in current function.

tagyieh committed:
Merge branch 'master' of https://github.com/Tobiky/mal-ls into feat/find_symbols_in_current_scope

Tobiky reviewed (changes requested):
While having additionally keyword symbols as part of the list works, as it can be used for e.g. the hover operation, users of the function will need a way to differentiate between them. For example, return two lists where one is user-defined symbols and keyword-symbols found.

Instead of using a list and filtering out items manually it may be better to initially use a set and then turn it into a list. More ergonomic and generally more performant (with the exception for very small lists).

Move the queries with only string literals outside the functions as well, so they're not reconstructed on every call. Since they are just strings, i.e. no formatting or string interpolation, we can construct them once the file is imported rather than every call.

The function should also not return keyword symbols that are not there.

Lastly, please associate them with their respective [cursor child index](https://tree-sitter.github.io/py-tree-sitter/classes/tree_sitter.TreeCursor.html#tree_sitter.TreeCursor.descendant_index) as mentioned in the issue.

tagyieh commented:
> The function should also not return keyword symbols that are not there.

For the `completion` functionality, couldn't it be useful for the user to receive all keywords, even those that aren't defined yet? For instance, to create an asset which extends another, while all other defined assets did not do this before

tagyieh commented:
> Lastly, please associate them with their respective [cursor child index](https://tree-sitter.github.io/py-tree-sitter/classes/tree_sitter.TreeCursor.html#tree_sitter.TreeCursor.descendant_index) as mentioned in the issue.

About this part, I am not sure how to do it. As of right now, we execute a query to search for the symbols, which returns a node. However, the attribute you indicated, `descendant_index`, is only available on cursors. This means that this attribute is only useful if we use the cursor itself to walk the tree and pattern match each node instead of using one query right away. In short, there is no relation between a cursor and the returned node from the query which allows us to use `descendant_index` for the desired purpose.  

An alternative would be to keep this as is, and in the parent issue we include if the index of parent, child, child of child, etc. via 1, -1, -2, etc. respectively. That is what I did, as initially I did not see that this issue also described this.

Do you have any input on this?

tagyieh committed:
fix: Changed to sets and moved queries.

Sets are now used to avoid duplicate items and constant queries were moved to the top of the file to only be built once.

Tobiky commented:
> > The function should also not return keyword symbols that are not there.
> 
> For the `completion` functionality, couldn't it be useful for the user to receive all keywords, even those that aren't defined yet? For instance, to create an asset which extends another, while all other defined assets did not do this before

While that is certainly good features for `completion`, that is not the responsibility/concern of this function.

> > Lastly, please associate them with their respective [cursor child index](https://tree-sitter.github.io/py-tree-sitter/classes/tree_sitter.TreeCursor.html#tree_sitter.TreeCursor.descendant_index) as mentioned in the issue.
> 
> About this part, I am not sure how to do it. As of right now, we execute a query to search for the symbols, which returns a node. However, the attribute you indicated, descendant_index, is only available on cursors. This means that this attribute is only useful if we use the cursor itself to walk the tree and pattern match each node instead of using one query right away. In short, there is no relation between a cursor and the returned node from the query which allows us to use descendant_index for the desired purpose.
> 
> An alternative would be to keep this as is, and in the parent issue we include if the index of parent, child, child of child, etc. via 1, -1, -2, etc. respectively. That is what I did, as initially I did not see that this issue also described this.
> 
> Do you have any input on this?
I must have misunderstood the docs when I had previously looked then. I guess instead we send back the node itself, which means we can skip the return of the symbol values/texts since that will be part of the node returned. The point is to allow for anyone using the function to be able to find where it originally was, perhaps for modification or finding other nearby information.

tagyieh commented:
> While that is certainly good features for `completion`, that is not the responsibility/concern of this function.

Understood
 
> I must have misunderstood the docs when I had previously looked then. I guess instead we send back the node itself, which means we can skip the return of the symbol values/texts since that will be part of the node returned. The point is to allow for anyone using the function to be able to find where it originally was, perhaps for modification or finding other nearby information.

So this function should only return two lists, one with user-defined strings and another with keywords?

tagyieh committed:
fix: Only return keywords present in node for category.

tagyieh committed:
fix: Only return keywords present in node for association.

tagyieh committed:
fix: Only return keywords present in node for asset.

Tobiky commented:
> > I must have misunderstood the docs when I had previously looked then. I guess instead we send back the node itself, which means we can skip the return of the symbol values/texts since that will be part of the node returned. The point is to allow for anyone using the function to be able to find where it originally was, perhaps for modification or finding other nearby information.
> 
> So this function should only return two lists, one with user-defined strings and another with keywords?

So there is no misunderstanding; not strings but the nodes themselves. If you have any other way of referencing the nodes then that is also fine, but then it should be the symbol value/string/text associated with the reference (tuple, dictionary, associative list, etc). The nodes are fine by themselves since the nodes will be the symbols and thus the value can be retrieved by fields/functions on the node. 

Otherwise, the separation between user-defined and keywords is entirely correct.

tagyieh commented:
> So there is no misunderstanding; not strings but the nodes themselves. If you have any other way of referencing the nodes then that is also fine, but then it should be the symbol value/string/text associated with the reference (tuple, dictionary, associative list, etc). The nodes are fine by themselves since the nodes will be the symbols and thus the value can be retrieved by fields/functions on the node.
> 
> Otherwise, the separation between user-defined and keywords is entirely correct.

In that case, I believe we can keep it as it is, since the user-defined symbols and keywords are now being separated and the keywords only include those which are already present

Tobiky commented:
@tagyieh I think there was a misunderstanding. If the strings are used to represent the symbols, then they have to have an associated reference back to the node in some capacity (i.e. where in the tree the symbol is) so it can be looked up separately if needed. Alternatively, nodes can be used by themselves since they already contain both parts of the information.

The returned lists needs to both say what the symbol is and where it is in the tree, through some way.

tagyieh commented:
@Tobiky Ah I see, sorry for the confusion! Thanks for the clarification, I will correct the lists to be dictionaries and include both the symbol and the corresponding node

tagyieh committed:
fix: Use dictionary to store node where symbols are found for category.

tagyieh committed:
fix: Use dictionary to store node where symbols are found for associations.

tagyieh committed:
fix: Use dictionary to store node where symbols are found for asset.

tagyieh commented:
@Tobiky I believe now the issue is fixed - a dictionary is returned, where keys are symbols and values are their node.

tagyieh committed:
fix: Use dictionary to store node where symbols are found for root node.

Tobiky commented:
@tagyieh One of the new tests are failing for me when I run it, specifically:
```
FAILED tests/integration/test_find_symbols_in_scope.py::test_find_symbols_in_root_node_scope - AssertionError: assert {'associations', 'category'} == {'association...gory', 'info'}
```

tagyieh commented:
> @tagyieh One of the new tests are failing for me when I run it, specifically:
> 
> ```
> FAILED tests/integration/test_find_symbols_in_scope.py::test_find_symbols_in_root_node_scope - AssertionError: assert {'associations', 'category'} == {'association...gory', 'info'}
> ```

Apologies, I forgot to upload the corrected test file. Now it should work

tagyieh committed:
fix: Upload correceted test file.

Tobiky reviewed (approved):
Splendid work, thanks for the patience
